api/src/main/java/com/epam/pipeline/manager/datastorage/providers/aws/s3/S3Helper.java

```diff
@@ -300,6 +303,44 @@ public void restoreFileVersion(String bucket, String path, String version) {
         moveS3Object(client, bucket, new MoveObjectRequest(path, version, path));
     }
 
+    public void restoreFolderVersion(String bucket, String path, RestoreFolderVO restoreFolderVO) {
+        AmazonS3 client = getDefaultS3Client();
+        try (S3ObjectDeleter deleter = new S3ObjectDeleter(client, bucket)) {
+            getDeletedItems(bucket, path, true, null, null, restoreFolderVO)
+                    .forEach(item -> deleter.deleteKey(item.getPath(), ((DataStorageFile) item).getVersion()));
+        } catch (SdkClientException e) {
+            throw new DataStorageException(e.getMessage(), e.getCause());
+        }
+    }
+
+    public List<AbstractDataStorageItem> getDeletedItems(String bucket, String path, Boolean showVersion,
+                                                         Integer pageSize, String marker, RestoreFolderVO restoreFolderVO) {
+        String requestPath = Optional.ofNullable(path).orElse("");
+        AmazonS3 client = getDefaultS3Client();
```

Comment:
Could you pass `client` trough method arguments? 
Also, make sure that new client does not created with recursion call (the idea is to create client only once)

Reply:
fixed